receiver/statsdreceiver/internal/protocol/statsd_parser.go

```diff
@@ -370,12 +376,87 @@ func (p *StatsDParser) Aggregate(line string, addr net.Addr) error {
 			// No action.
 		}
 	}
+}
+
+// statsDMetricIter allows alloc-free iteration over statsDMetric values.
+// This is helpful to support multiple values in a single statsd metric
+// such as timers, etc.
+type statsDMetricIter struct {
+	statsDMetricProperties
+
+	// valueStr contains potentially many values separated by ":".
+	valueStr string
+}
+
+type statsDMetricProperties struct {
+	description statsDMetricDescription
+	unit        string
+	sampleRate  float64
+	timestamp   uint64
+}
+
+func (i *statsDMetricIter) All() ([]statsDMetric, error) {
+	res := make([]statsDMetric, 0, strings.Count(i.valueStr, ":")+1)
+	if err := i.ForEachValue(func(m statsDMetric) {
+		res = append(res, m)
+	}); err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (i *statsDMetricIter) ForEachValue(f func(m statsDMetric)) error {
```

Comment:
nit: Maybe add a reason why both All and ForEachValue exist? (call out that foreachvalue avoids heap allocs?)